ngramoffset.go

```diff
@@ -112,3 +122,211 @@ func (a *arrayNgramOffset) DumpMap() map[ngram]simpleSection {
+				ngramsAscii = append(ngramsAscii, ngp, ngp)
+				offsetsAscii = append(offsetsAscii, offsets[i], offsets[i+1])
+			} else {
+				ngramsAscii = append(ngramsAscii, ngp)
+				offsetsAscii = append(offsetsAscii, offsets[i])
+			}
+			if offsets[i+1]-offsets[i] >= ngramAsciiMaxSectionLength {
+				// max-length ascii sections can't be represented properly in the ascii mapping,
+				// and are duplicated in the normal unicode entries.
+				ngramsUnicode = append(ngramsUnicode, ng, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i], offsets[i+1])
+			}
+		} else {
+			if i == len(ngrams)-1 || ngrams[i+1]&ngramAsciiMask == ngrams[i+1] {
+				ngramsUnicode = append(ngramsUnicode, ng, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i], offsets[i+1])
+			} else {
+				ngramsUnicode = append(ngramsUnicode, ng)
+				offsetsUnicode = append(offsetsUnicode, offsets[i])
+			}
+		}
+	}
+
+	if len(ngramsAscii) > 0 {
+		ngramsAscii = ngramsAscii[:len(ngramsAscii)-1]
+	}
+	if len(ngramsUnicode) > 0 {
+		ngramsUnicode = ngramsUnicode[:len(ngramsUnicode)-1]
+	}
+
+	asc := makeAsciiNgramOffset(ngramsAscii, offsetsAscii)
+	uni := makeArrayNgramOffset(ngramsUnicode, offsetsUnicode)
+
+	return combinedNgramOffset{asc, &uni}
+}
+
+// Get returns a simpleSection with sz=0 if no entry, or sz=ngramAsciiMaxSectionLength
+// if the length of the ngram is too large for this type and it should cascade to the next entry.
+func (a combinedNgramOffset) Get(gram ngram) simpleSection {
+	if a.asc == nil {
+		return simpleSection{}
+	}
+
+	var sec simpleSection
+	if gram&ngramAsciiMask == gram {
+		sec = a.asc.Get(gram)
+	}
+	if sec.sz == 0 || sec.sz == ngramAsciiMaxSectionLength {
+		sec = a.uni.Get(gram)
+	}
+	return sec
+}
+
+func (a combinedNgramOffset) DumpMap() map[ngram]simpleSection {
+	m := a.asc.DumpMap()
+	for k, v := range a.uni.DumpMap() {
+		m[k] = v
+	}
+	return m
+}
+
+func (a combinedNgramOffset) SizeBytes() int {
+	return a.asc.SizeBytes() + a.uni.SizeBytes()
+}
+
+const ngramAsciiMask = 127 | 127<<21 | 127<<42
+
+// Ascii mapping packs 3*7b chars and 11 bits of lengths, with this as the set maximum.
+// We could save another ~3% of total RAM / 5% of combinedNgramOffset RAM by switching to
+// a 40b packing with 19-bit lengths, but the code would be significantly uglier so it doesn't
+// seem worth it.
+const ngramAsciiMaxSectionLength = (1 << 11) - 1
+
+func ngramAsciiToPacked(ng ngram) uint32 {
+	return uint32(ng&127) | uint32((ng>>(21-7))&(127<<7)) | uint32((ng>>(42-14))&(127<<14))
+}
+
+func ngramAsciiPackedToNgram(ng uint32) ngram {
+	return ngram(ng&127) | ngram(ng&(127<<7))<<(21-7) | ngram(ng&(127<<14))<<(42-14)
+}
+
+// asciiNgramOffset stores ascii trigrams packed together with short lengths,
+// using offsets for a chunk of entries to limit the number of lengths that must
+// be summed to compute a section's offset.
+type asciiNgramOffset struct {
+	entries      []uint32 // (chara << 25 | charb << 18 | charc << 11 | length)
```

Comment:
I'm unsure about this idea, so just gonna float it to see what you think. A `type asciiNgramLenPacked uint32` which has methods to return the ngram and length. The type will also help us know when we are dealing with the packed type. Probably a terrible name, but more about the idea of a type for the packed data.